src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -31,6 +32,8 @@ internal sealed partial class SolutionCompilationState
     /// </summary>
     private static readonly ConditionalWeakTable<ISymbol, ProjectId> s_assemblyOrModuleSymbolToProjectMap = new();
 
+    private static readonly ObjectPool<SegmentedDictionary<ProjectId, ICompilationTracker>> s_trackerInfoPool = new(() => new());
```

Comment:
❓ Can you run another pass through without this pool? Instead of using an intermediate dictionary, just use `_projectIdToTrackerMap.ToBuilder()` at the start of the method, and then return `builder.ToImmutable()` at the end. This will always create a small builder object, but it removes the need for `allReused` and `isModified`, and only allocates a new dictionary if/when something changes.

Reply:
Sure, I can give that a go, results probably won't come out for around 12 hours. 

From my debugging, I saw a large number of calls during project open come through where projectIdToTrackerMap was empty, so I don't love that this would allocate, but we'll give it a go.

Reply:
> From my debugging, I saw a large number of calls during project open come through where projectIdToTrackerMap was empty

Thanks, this is very helpful. I reviewed all 5 call sites based on this, and found:

* 3 call sites never add new items to the map, and would silently proceed if the map was empty. These call sites can pass `skipEmpty: true` (name could be changed), allowing the call to return immediately if the map is empty.
* 1 call site (`WithFrozenSourceGeneratedDocuments`) will add items to the map in the callback. This call site can pass `skipEmpty: false`.
* 1 call site (`WithoutFrozenSourceGeneratedDocuments`) does not add items to the map, but it has a runtime assertion that the map is not empty in the callback. This call site can also pass `skipEmpty: false` to preserve the assertion behavior.